src/wui/mapview.cc

```diff
@@ -587,10 +586,56 @@ Widelands::NodeAndTriangle<> MapView::track_sel(const Vector2i& p) {
 	return node_and_triangle;
 }
 
+bool MapView::scroll_map() {
+	// arrow keys
+	const bool kUP = get_key_state(SDL_SCANCODE_UP);
+	const bool kDOWN = get_key_state(SDL_SCANCODE_DOWN);
+	const bool kLEFT = get_key_state(SDL_SCANCODE_LEFT);
+	const bool kRIGHT = get_key_state(SDL_SCANCODE_RIGHT);
+
+	// numpad keys
+	const bool kNumlockOff = !(SDL_GetModState() & KMOD_NUM);
+	const bool kNP1 = get_key_state(SDL_SCANCODE_KP_1);
+	const bool kNP2 = get_key_state(SDL_SCANCODE_KP_2);
+	const bool kNP3 = get_key_state(SDL_SCANCODE_KP_3);
+	const bool kNP4 = get_key_state(SDL_SCANCODE_KP_4);
+	const bool kNP6 = get_key_state(SDL_SCANCODE_KP_6);
+	const bool kNP7 = get_key_state(SDL_SCANCODE_KP_7);
+	const bool kNP8 = get_key_state(SDL_SCANCODE_KP_8);
+	const bool kNP9 = get_key_state(SDL_SCANCODE_KP_9);
+
+	// set the scrolling distance
+	uint8_t denominator =
+	   ((SDL_GetModState() & KMOD_SHIFT) ? 4 : (SDL_GetModState() & KMOD_CTRL) ? 16 : 8);
+	uint16_t scroll_distance_y = g_gr->get_yres() / denominator;
+	uint16_t scroll_distance_x = g_gr->get_xres() / denominator;
+	int32_t distance_to_scroll_x = 0;
+	int32_t distance_to_scroll_y = 0;
+
+	// check the directions
+	if (kUP || (kNumlockOff && (kNP7 || kNP8 || kNP9)))
+		distance_to_scroll_y -= scroll_distance_y;
+	if (kDOWN || (kNumlockOff && (kNP1 || kNP2 || kNP3)))
+		distance_to_scroll_y += scroll_distance_y;
+	if (kLEFT || (kNumlockOff && (kNP1 || kNP4 || kNP7)))
+		distance_to_scroll_x -= scroll_distance_x;
+	if (kRIGHT || (kNumlockOff && (kNP3 || kNP6 || kNP9)))
+		distance_to_scroll_x += scroll_distance_x;
+
+	// do the actual scrolling
+	if (distance_to_scroll_x == 0 && distance_to_scroll_y == 0)
+		return false;
+	pan_by(Vector2i(distance_to_scroll_x, distance_to_scroll_y), Transition::Smooth);
+	return true;
+}
+
```

Comment:
Not needed, but I wouldn't omit the block brackets on ifs, as its then easier to add code, etc.